#+title: Cross-symbol checker

This tool maps aliases, previous or withdrawn symbols to current approved HGNC symbols. Checks for any entries that are not gene symbols and fixes capitalizations. Ensembl and RefSeq annotation files sometimes use previous or alias symbols. This tool also checks Ensembl and NCBI annotation files for different genome versions and shows which gene symbol is used. This way a more appropriate gene set can be used to avoid the false negatives in the variant discovery process.

* How to install

** Install from git
Clone this repository to install.

~git clone https://github.com/barslmn/bioscripts~

To get the most up to date annotation files run the ~./get-data.sh~. This process takes some time as it downloads various gene symbol annotation files from NCBI and Ensembl and HGNC. Check install section at [[file:genesymbolchecker.org][cross-symbol-checker.org]] to see what is being downloaded.

** Get container
Installing with docker.

~docker pull barslmn/cross-symbol-checker~

* How to use
Main operation is performed at ~cross-symbol-checker.sh~ this file takes a single symbol as its input and returns the status of all of the approved, alias, previous or withdrawn symbols in all available annotation sources.
Another script is ~check-geneset.sh~ which takes one or more gene and target assembly as its arguments. This script is a wrapper around cross-symbol-checker and parses its output.

*Example use:*
Check single gene for all annotation sources:
~cross-symbol-checker.sh SHFM6~

Check multiple genes for given annotation source:
~check-geneset.sh SHFM6 ADA2 -a GRCh37 -s RefSeq~

*Example use with docker:*
Check single gene for all annotation sources:
~docker run barslmn:cross-symbol-checker /opt/cross-symbol-checker/cross-symbol-checker.sh SHFM6~

Check multiple genes for given annotation source:
~docker run barslmn:cross-symbol-checker /opt/cross-symbol-checker/check-geneset.sh SHFM6 ADA2 -a GRCh37 -s RefSeq~
